### PR TITLE
feat(gatsby-cli): Add error codes and structured errors

### PR DESCRIPTION
**Pull request by m-allanson**
## Description

This adds a dedicated error component to the cli, and the start of an error map which lets us identify individual error messages and eventually add better and more consistent info to all errors.

Quick notes:

Updates before merging:

- [x] this should be hooked up to report telemetry about errors
- [x] there are several ad-hoc ways of displaying the original error message (origError, sourceMessage), standardise on one way of doing this
- [x] Update from babelCodeFrame to codeFrameColumns is incomplete (check lint errors). 
- [x] In file-parser, need to add `templateLoc` to StaticQuery as well

Future improvements:
- [x] fix `[Object object]` output in stacks for html build errors `WebpackError: ReferenceError: window is not defined\n    at Header (src/components/header.js:6:15)\n    at [object Object]\n    at [object Object]' `
- Handle multipleRootQueries error
- In query-compiler, ‘reportError’ is used to display the browser overlay error screen in the browser. Ensure that report.error is used to report the same error to console
- use new structured errors to handle browser overlay errors, so we can report errors in the same way
- render and display code frames 
- update cleanPaths to handle null
- ensure we create rich errors from [api-runner-node](https://github.com/gatsbyjs/gatsby/blob/ddb007852ac7091c881c516fa2b6473f5343a6eb/packages/gatsby/src/utils/api-runner-node.js#L209-L213)
- MDX?! :D

### CODE REVIEW

**Comment by m-allanson** [comment]
Working through some message improvements. Small changes so far, but validating that this approach will work for a variety of messages.

Before:

Screenshot 2019-06-20 at 13 23 22

After:

Screenshot 2019-06-20 at 13 41 11

**Comment by DSchau** [comment]
@m-allanson that is looking sick! Nicely done!

Presuming we'll eventually link to documentation from the errors, too?

**Comment by m-allanson** [comment]
> Presuming we'll eventually link to documentation from the errors, too?

@DSchau yes! We've made a schema for errors that has space for a docsUrl so each error can (optionally) link to a specific docs page.

File: packages/gatsby-telemetry/src/telemetry.js [hunk 1]
```diff
@@ -106,13 +106,32 @@ module.exports = class AnalyticsTracker {
...
+        return
+      }
+
+      tags.error_v2 = {
+        id: tags.error.id,
+        text: cleanPaths(tags.error.text),
+        level: tags.error.level,
+        type: tags.error?.type,
+        // see if we need empty string or can just use NULL
+        stack: cleanPaths(tags.error?.error?.stack || ``),
```
Comment (jamo) [inline packages/gatsby-telemetry/src/telemetry.js]: We can fix the cleanPaths to handle `null`

**Pull request merged**